rust/lance-index/src/vector/ivf/shuffler.rs

```diff
@@ -95,69 +260,84 @@ pub async fn shuffle_dataset(
-                        let row_ids = batch.column_by_name(ROW_ID).ok_or(Error::Index {
-                            message: "column does not exist".to_string(),
-                            location: location!(),
-                        })?;
-                        let part_ids = UInt32Array::from_iter(
-                            row_ids
-                                .as_primitive::<UInt64Type>()
-                                .values()
-                                .iter()
-                                .map(|row_id| partition_map.get(row_id).copied()),
-                        );
-                        let part_ids = UInt32Array::from(part_ids);
-                        batch = batch
-                            .try_with_column(
-                                Field::new(PART_ID_COLUMN, part_ids.data_type().clone(), true),
-                                Arc::new(part_ids.clone()),
-                            )
-                            .expect("failed to add part id column");
-
-                        if part_ids.null_count() > 0 {
-                            info!(
-                                "Filter out rows without valid partition IDs: null_count={}",
-                                part_ids.null_count()
-                            );
-                            let indices = UInt32Array::from_iter(
-                                part_ids
-                                    .iter()
-                                    .enumerate()
-                                    .filter_map(|(idx, v)| v.map(|_| idx as u32)),
+        let stream =
+            data.zip(repeat_with(move || ivf.clone()))
+                .map(move |(b, ivf)| {
+                    // If precomputed_partitions map is provided, use it
+                    // for fast partitions.
+                    let partition_map = precomputed_partitions
+                        .as_ref()
+                        .cloned()
+                        .unwrap_or(Arc::new(Vec::new()));
+                    let nan_filter = KeepFiniteVectors::new(&column);
+
+                    tokio::task::spawn(async move {
+                        let mut batch = b?;
+
+                        if !partition_map.is_empty() {
+                            let row_ids = batch.column_by_name(ROW_ID).ok_or(Error::Index {
+                                message: "column does not exist".to_string(),
+                                location: location!(),
+                            })?;
+                            let part_ids = UInt32Array::from_iter(
+                                row_ids.as_primitive::<UInt64Type>().values().iter().map(
+                                    |row_id| {
+                                        let row_addr = RowAddress::new_from_id(*row_id);
+                                        partition_map[row_addr.fragment_id() as usize]
+                                            [row_addr.row_id() as usize]
```

Comment:
Are you re-using the row id concept but in index files? I do worry it could be confusing.

Reply:
Index files need to refer to something, either row id or row address.  So the problem statement here is that we have a GPU process (possibly distributed on multiple systems in the future) that calculates the partition id for each row id.  We then need to pass this mapping forward.

We could use `HashMap<u64, u32>` (we did originally) but this gets very expensive at high scales (in the future, if we distribute this task as well, maybe not the end of the world).

So basically it is very nice if there is some reliable mechanism to create a "row identifier" -> X mapping without using a hashmap.  With row addresses we can do this easily using two `Vec` and it should be stable as long as the read version used to create the mapping is the same as the read version used to apply the mapping (it is).

If we wanted to use row ids here then we could maybe assume that the number of "gaps" will be rather small and then just use a straight `Vec` (one advantage is now it doesn't need to be the same read version but we might have missing entries otherwise) but I don't know if we can always guarantee these small gaps and, if there are large gaps, we end up allocating more RAM than strictly needed.  What do you think?

Worst case we can just assume that stage 2 is always run on the same data, in the same order, as stage 1.  Then we can just do a list of partition ids.  I think this might not work today because we process things in parallel and don't order them (but I don't think the cost of doing that ordering would be significant compared to everything else we do).

Reply:
Okay this makes sense. I think I was worried that you were re-using the `RowAddress` idea for something that isn't a row address. It looked at first like it was the address of the vector within the index partitions. If it's actually RowAddresses as we know now, I agree this makes sense.